Reproducer should deduce artifact root from concrete environment

kwryankrattiger opened the pull request:
<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

Fix an bug in the reproducer that hard coded the artifacts root directory.

CC @scottwittenburg @zackgalbreath

kwryankrattiger committed:
Reproducer should decude artifact root from concrete environment

kwryankrattiger commented:
Confirmed the reproducer still works using https://gitlab.spack.io/spack/spack/-/jobs/11963223#L980

scheibelp commented:
Close/reopen to restart tests

johnwparent reviewed (dismissed):
Looks good to me!

scheibelp commented on lib/spack/spack/ci.py:
```diff
@@ -1530,10 +1530,20 @@ def download_and_extract_artifacts(url, work_dir):
 
     os.remove(artifacts_zip_path)
 
+    return artifact_root
```
Can you update docstring with the return type/info?

Or perhaps if it refers to whatever description there is of the overall layout of this directory? I did not find one briefly looking through here and `cmd/ci.py`: I would have expected to see something about it in `generate_gitlab_ci_yaml` (which seems largely responsible for assembling the reproducer tarball). If that in fact does not exist, would you be able to add that as well?

kwryankrattiger replied:
I can update the doc string.

I also added some docs on the artifacts root layout. I couldn't find anything else describing it and it is probably useful to have written down somewhere.

scheibelp commented on lib/spack/spack/test/cmd/ci.py:
```diff
@@ -210,7 +210,9 @@ def test_ci_generate_with_env(
-            ci_cmd("generate", "--output-file", outputfile)
+            ci_cmd(
+                "generate", "--artifacts-root", "my_artifacts_root", "--output-file", outputfile
+            )
```
(no change needed - question) If I understand: the only time an end user would see this problem is if they used `--artifacts-root`, which otherwise defaults to `"jobs_scratch_dir"`, and they would encounter this on the "consumer end" (`spack ci reproduce-build ...`). Also it looks like this will work with reproducer archives generated before this PR.

My primary reason for asking is to make sure I can generate a sensible commit message.

kwryankrattiger replied:
That is correct. This is just a fix for the consumer end of the reproducer and does not change the the layout of job artifacts.

scheibelp reviewed (dismissed):
I have one documentation request and one question

kwryankrattiger committed:
Add documentation on the layout of the artifacts directory

kwryankrattiger commented on lib/spack/docs/pipelines.rst:
```diff
@@ -754,6 +754,69 @@ presence of a ``SPACK_CDASH_AUTH_TOKEN`` environment variable during the
 
+.. _ci_artifacts:
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
@scottwittenburg if you have anything to add from historical context here as well that would be helpful. I mostly gathered this from what the code seemed to be doing.

scottwittenburg commented on lib/spack/docs/pipelines.rst:
```diff
@@ -754,6 +754,69 @@ presence of a ``SPACK_CDASH_AUTH_TOKEN`` environment variable during the
+CI Artifacts Directory Layout
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When running the CI build using the command ``spack ci rebuild`` a number of direcotries are created for
```
```suggestion
When running the CI build using the command ``spack ci rebuild`` a number of directories are created for
```

scottwittenburg reviewed (dismissed):
Thanks @kwryankrattiger!  I just noticed one typo, otherwise it looks good to me.

kwryankrattiger commented:
waiting to rebase on https://github.com/spack/spack/pull/47459

kwryankrattiger committed:
Use dag hash in the container name

kwryankrattiger committed:
Add reproducer options to improve local testing

* --overwrite allows reusing reproducer directory
* --use-local-head allows running reproducer with
  the current Spack HEAD commit rather than computing
  a commit for the reproducer

kwryankrattiger committed:
Add test to verify commits and recreating reproduction environment

kwryankrattiger committed:
Add test for non-merge commit case

scottwittenburg commented:
I tried to reproduce a `develop` [failure](https://gitlab.spack.io/spack/spack/-/jobs/15087851) using this branch.  Maybe I just took the perfectly wrong set of steps: first I tried just pasting the command from the trace (without any extra arguments), and the reproducer failed to verify signed binaries.  So I tried again with `--gpg-url` and got:

<details>
  <summary>stack trace</summary>
```
$ spack ci reproduce-build --gpg-url https://spack.github.io/keys/spack-public-binary-key.pub https://gitlab.spack.io/api/v4/projects/2/jobs/15087851/artifacts
==> Fetching artifacts from: https://gitlab.spack.io/api/v4/projects/2/jobs/15087851/artifacts
==> Error: error parsing YAML: near /home/local/KHQ/scott.wittenburg/projects/spack/spack/ci_reproduction/spack/lib/spack/spack/test/data/config/compilers.yaml, 2, 4: expected <block end>, but found '{'
while parsing a block mapping
  in "/home/local/KHQ/scott.wittenburg/projects/spack/spack/ci_reproduction/spack/lib/spack/spack/test/data/config/compilers.yaml", line 3, column 5
expected <block end>, but found '{'
  in "/home/local/KHQ/scott.wittenburg/projects/spack/spack/ci_reproduction/spack/lib/spack/spack/test/data/config/compilers.yaml", line 4, column 38
```
</details>

Hard to know what to make of that.  I assumed maybe it was because it tried to use the same default working directory.  So I provided the new `--overwrite` argument and got:

```
$ spack ci reproduce-build --overwrite --gpg-url https://spack.github.io/keys/spack-public-binary-key.pub https://gitlab.spack.io/api/v4/projects/2/jobs/15087851/artifacts
==> Error: [Errno 13] Permission denied: 'package.cpython-36.pyc'
```

Any suggestions?

scottwittenburg commented:
Yeah, using `sudo` works, as you'd expect.  Also, manually removing the working dir (again, with `sudo`, as expected).  So I'm not sure if `--overwrite` is just missing a line of documentation to be useful, or if it's really worth it.  If you want to keep it, should the first error be improved so people know what's wrong? I just guessed it was from re-using the default working dir.

Regarding the `--use-local-head` argument, I wonder what are the cases where that is useful.  If fixing the error which is being reproduced requires a change to a `package.py`, can `--use-local-head` actually help? Or really, can it help with anything that requires a re-concretization?

kwryankrattiger commented:
>  So I'm not sure if --overwrite is just missing a line of documentation to be useful, or if it's really worth it. If you want to keep it, should the first error be improved so people know what's wrong? I just guessed it was from re-using the default working dir.

It can be useful for testing and cases where the command fails at some step before the container runs. You can make it work all the time by running the reproduce build as `sudo`...but ymmv on that one.

> Regarding the --use-local-head argument, I wonder what are the cases where that is useful.

Little known fact about Spack environments, if you have a concrete environment, and you do not re-concretize explicitly, you can modify the `package.py` of the package you are working on and run the new recipe using the old hash. This is, in my opinion, a feature not a bug. Granted, I don't know if it is intentional or not.

scottwittenburg commented:
> You can make it work all the time by running the reproduce build as `sudo`

Does that deserve a note somewhere then?  Maybe in the `--help` output, or elsewhere?

And it still seems to me that `error parsing YAML` in a stack trace doesn't help the user to know that they need to remove the working directory.

> Little known fact about Spack environments, if you have a concrete environment, and you do not re-concretize explicitly, you can modify the package.py of the package you are working on and run the new recipe using the old hash.

TIL.  And in that case, I agree that `--use-local-head` could be nice.

scottwittenburg commented:
Neither of those seem particularly helpful in solving the actual problem either.  Should `spack ci reproduce-build` check for a non-empty working directory, and if that is the case, print an error saying the user needs to point to an empty directory?

kwryankrattiger commented:
> Neither of those seem particularly helpful in solving the actual problem either. Should `spack ci reproduce-build` check for a non-empty working directory, and if that is the case, print an error saying the user needs to point to an empty directory?

That seems pretty reasonable, the biggest issue I am trying to solve is running reproducer on a non-directory will almost always fail.

kwryankrattiger committed:
ci reproduce-build: Drop overwrite option
in favor of throwing an error if the working dir is non-empty

kwryankrattiger committed:
Update completions

scottwittenburg commented on lib/spack/spack/ci/__init__.py:
```diff
@@ -637,13 +641,25 @@ def download_and_extract_artifacts(url, work_dir):
+            artifact_root = ""
+            for f in zip_file.filelist:
+                if "spack.lock" in f.filename:
+                    artifact_root = os.path.dirname(os.path.dirname(f.filename))
```
It was a pain to trace through from here looking for where this value is born, I'm sorry we both had to do that :laughing:  I know there is still lots to do to make the ci modules nicer to work with.

scottwittenburg reviewed (approved):
Thanks for these improvements @kwryankrattiger!
